### PR TITLE
Prototype selection destructive maxdist

### PR DESCRIPTION
**Pull request by sjanssen2**
Most promising heuristic for prototype selection

### CODE REVIEW

File: phylogeny/prototypeSelection.py [hunk 1]
```diff
@@ -369,3 +371,161 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
...
+    currDists[minElmIdx], numRemain = np.infty, numRemain-1
+
+    # continue until only num_prototype elements are left
+    while (numRemain > num_prototypes):
+        # substract the distance to the removed element for all remaining
+        # elements
+        currDists -= dm.data[minElmIdx]
+        # find the next element to be removed, again as the one that is
+        # clostest to all others
+        minElmIdx = currDists.argmin()
```
Comment (qiyunzhu) [inline phylogeny/prototypeSelection.py:524]: "clostest" is a typo...

**Review by qiyunzhu — COMMENTED** [review]
Good!

**Comment by qiyunzhu** [comment]
This one will need some review too @wasade @antgonza @ekopylova if you would like.

**Comment by wasade** [comment]
@sjanssen2 would it be possible to pull from upstream master? I think commits in here overlap with #46

File: phylogeny/prototypeSelection.py (hunk 1 above)
Comment (sjanssen2) [reply phylogeny/prototypeSelection.py:524]: fixed. Thanks!

**Comment by sjanssen2** [comment]
Merged with biocore master and fixed one typo.

**Comment by qiyunzhu** [comment]
@sjanssen2 Thanks for merging! @wasade

File: phylogeny/prototypeSelection.py [hunk 2]
```diff
@@ -448,3 +449,83 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
         prototypes.append(np.asarray(minVals).argmin())
 
     return [dm.ids[idx] for idx in prototypes]
+
+
+def prototype_selection_destructive_maxdist(dm, num_prototypes):
+    '''Heuristically select k prototypes for given distance matrix.
+
+       Prototype selection is NP-hard. This is an implementation of a greedy
+       correctness heuristic: Start with the complete set and interatively
```
Comment (wasade) [inline phylogeny/prototypeSelection.py]: interatively -> iteratively

File: phylogeny/prototypeSelection.py [hunk 3]
```diff
@@ -448,3 +449,83 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
...
+    currDists = dm.data.sum(axis=1)
+
+    # the element to remove first is the one that has smallest distance to all
+    # other. "Removing" works by tagging its distance-sum as infinity. Plus, we
+    # decrease the number of available elements by one.
+    minElmIdx = currDists.argmin()
+    currDists[minElmIdx], numRemain = np.infty, numRemain-1
+
+    # continue until only num_prototype elements are left
+    while (numRemain > num_prototypes):
```
Comment (wasade) [inline phylogeny/prototypeSelection.py:518]: It seems like this block of code is effectively:

```
np.argsort(currDists)[:num_prototypes]
```

...unless I'm missing something? Won't this always find the next smallest distance sum, which in effect would be the above `argsort`? I don't see how the rank order of the values would change given the while block.

File: phylogeny/prototypeSelection.py [hunk 4]
```diff
@@ -448,3 +449,83 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
...
+    num_prototypes: int
+        Number of prototypes to select for distance matrix.
+        Must be >= 2, since a single prototype is useless.
+        Must be smaller than the number of elements in the distance matrix,
+        otherwise no reduction is necessary.
+
+    Returns
+    -------
+    list of str
+        A sequence holding selected prototypes, i.e. a sub-set of the
```
Comment (wasade) [inline phylogeny/prototypeSelection.py:476]: From the return, it looks like this is a subset of the IDs of the matrix not the elements?

File: phylogeny/prototypeSelection.py (hunk 2 above)
Comment (sjanssen2) [reply phylogeny/prototypeSelection.py]: fixed

File: phylogeny/prototypeSelection.py (hunk 4 above)
Comment (sjanssen2) [reply phylogeny/prototypeSelection.py:476]: you are right. I fixed the documentation (of all heuristics)

File: phylogeny/prototypeSelection.py (hunk 3 above)
Comment (sjanssen2) [reply phylogeny/prototypeSelection.py:518]: I tend to say that you miss something here. It is more than sorting a list and taking the first X elements. In each iteration, some list elements change and thus the ordering is effected: `currDists -= dm.data[minElmIdx]`
I'd be happy to revise the whole loop with a single line of code, but I don't see how.

Comment (wasade) [reply phylogeny/prototypeSelection.py:518]: Oh, I see it now. Sorry for the confusion.

**Comment by sjanssen2** [comment]
OK to merge?

**Pull request merged**